src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivations (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(M : Type*) [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace instances
+def smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := infer_instance
+attribute [instance, priority 10000] smooth_functions_algebra
+def tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := infer_instance
+attribute [instance, priority 10000] tower
+def sizeof : has_sizeof (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^⊤⟮I, M; 𝕜⟯) := infer_instance
+attribute [instance, priority 100000] sizeof
+end instances
+
+namespace point_derivation
+
+/-- The scalar multiplication depends on the point `x : M`. -/
+def has_scalar (x : M) : has_scalar C^∞⟮I, M; 𝕜⟯ 𝕜 :=
+{ smul := λ f k, f x * k }
+
+lemma scalar_def {x : M} {f : C^∞⟮I, M; 𝕜⟯} {k : 𝕜} :
+  @has_scalar.smul C^∞⟮I, M; 𝕜⟯ 𝕜 (has_scalar I M x) f k = f x * k := rfl
+
+/-- The scalar multiplication defined above gives rise to a module structure. -/
+def module (x : M) : module C^∞⟮I, M; 𝕜⟯ 𝕜 :=
```

Comment:
I think it might be better to factor this through `module.comp_hom`. E.g.,
```lean
/-- Perhaps this already exists? -/
def smooth_function.eval (x : M) : C^∞⟮I, M; 𝕜⟯ →ₐ[𝕜] 𝕜 :=
{ to_fun    := λ f, f x,
  map_one'  := sorry,
  map_mul'  := sorry,
  map_zero' := sorry,
  map_add'  := sorry,
  commutes' := sorry, }
```
and then something like
```lean
def module (x : M) : module C^∞⟮I, M; 𝕜⟯ 𝕜 :=
module.comp_hom 𝕜 (smooth_function.eval I M x).to_ring_hom
```

Reply:
This looks cool! But you say this because it is more performant or because `smooth_function.eval` could come in useful later?

Reply:
I mean the fact that the evalutation at a point behaves as an algebra morphism is already taken care in the `pi` namespace for maps that are valued in algebras, although the coercion is not registered as an algebra morphism I guess

Reply:
I make no claims about performance. It's just that mathematically this is how I think about this.

(In fact mathematically, the way I _really_ think about it is via first defining:
```lean
include I
def scalar_bundle := Σ x : M, 𝕜
```
and then defining the module strutures on the fibers of `scalar_bundle I M` but I'm not sure if that helps here.)

Reply:
And of course, yes: I expect we'll want `smooth_function.eval` sooner or later, regardless of whether we use it here.

Reply:
We should probably have `smooth_function.eval_alg_hom`, `smooth_function.eval_ring_hom`, ..., just as we have `pi.eval_alg_hom`, `pi.eval_ring_hom`, ....

These are easiest to build out of `smooth_function.coe_alg_hom` composed with `pi.alg_hom`.

Reply:
But why so? I mean why doing it here and not for any other object that has a coercion to a function? If it is only for the definition of algebraic structures they can also be defined directly as I did. I mean mathlib does already have many lemmas that deal with functions valued in an algebraic structure that do not go through registering evaluation as a morphism of that given structure, so I don't see what makes this situation different than the rest of mathlib!

Reply:
We do have these for other objects with function coercions though; see `monoid_hom.eval`, ` continuous_map.coe_fn_monoid_hom`, `quadratic_form.eval_add_monoid_hom`, `linear_map.applyₗ`, etc.

I added some more of these in #7720, but I now realize that was `continuous_map` not `smooth_function`!

Reply:
I implemented this definition of the module structure although as you see it triggers a deterministic timeout for the following lemma:
```
@[simp] lemma fdifferential_comp (g : C^∞⟮I', M'; I'', M''⟯) (f : C^∞⟮I, M; I', M'⟯) (x : M) :
  (𝒅g (f x)).comp (𝒅f x) = 𝒅(g.comp f) x :=
by { ext, simp only [apply_fdifferential], refl }
```
Which on my local machine compiles but takes several minutes... I will try to investigate why

Reply:
It is again a subsingleton problem it seems...

Reply:
Or give me some hints about what try to do further, other than using the profiler and tracing the type class inference I don't know what else to try...

Reply:
Actually I found the solution: flipping the equality solves all problems...